flink-runtime/src/test/java/org/apache/flink/runtime/security/FlinkUserSecurityManagerTest.java

```diff
@@ -0,0 +1,187 @@
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests for {@code FlinkUserSecurityManager}.
+ */
+public class FlinkUserSecurityManagerTest extend TestLogger {
+
+	private static final int TEST_EXIT_CODE = 123;
+	private FlinkUserSecurityManager flinkUserSecurityManager;
+
+	@Before
+	public void setUp() {
+		flinkUserSecurityManager = new FlinkUserSecurityManager(FlinkUserSecurityManager.CheckExitMode.THROW);
+		System.setSecurityManager(flinkUserSecurityManager);
+	}
+
+	@After
+	public void tearDown() {
+		if (flinkUserSecurityManager != null) {
+			System.setSecurityManager(flinkUserSecurityManager.getOriginalSecurityManager());
+		}
+	}
+
+	@Test(expected = UserSystemExitException.class)
+	public void testThrowUserExit() {
+		flinkUserSecurityManager.monitorSystemExit();
+		flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+	}
+
+	@Test
+	public void testToggleUserExit() {
+		flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+		flinkUserSecurityManager.monitorSystemExit();
+		try {
+			flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+			fail();
+		} catch (UserSystemExitException ignored) { }
+		flinkUserSecurityManager.unmonitorSystemExit();
+		flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+	}
+
+	@Test
+	public void testPerThreadThrowUserExit() throws Exception {
+		ExecutorService executorService = Executors.newSingleThreadExecutor();
+		// Async thread test before enabling monitoring ensures it does not throw while prestarting
+		// worker thread, which is to be unmonitored and tested after enabling monitoring enabled.
+		CompletableFuture<Void> future =
+			CompletableFuture.runAsync(() -> flinkUserSecurityManager.checkExit(TEST_EXIT_CODE),
+				executorService);
+		future.get();
+		flinkUserSecurityManager.monitorSystemExit();
+		try {
+			flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+			fail();
+		} catch (UserSystemExitException ignored) { }
+		// This threaded exit should be allowed as thread is not spawned while monitor is enabled.
+		future = CompletableFuture.runAsync(() -> flinkUserSecurityManager.checkExit(TEST_EXIT_CODE),
+				executorService);
+		future.get();
+	}
+
+	@Test
+	public void testInheritedThrowUserExit() throws Exception {
+		flinkUserSecurityManager.monitorSystemExit();
+		try {
+			flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+			fail();
+		} catch (UserSystemExitException ignored) { }
+		Thread thread = new Thread(() -> {
+			try {
+				flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+				fail();
+			} catch (UserSystemExitException ignored) {
+			} catch (Throwable t) {
+				fail();
```

Comment:
This `fail()` won't fail the test, as it is thrown from the thread. JUnit only fails a test if the main thread fails.

Maybe there's a nicer solution, but that's what I came up with:
```java
    @Test
    public void testInheritedThrowUserExit() throws Throwable {
        flinkUserSecurityManager.monitorSystemExit();
        try {
            flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
            fail();
        } catch (UserSystemExitException ignored) {
        }
        Tuple1<Throwable> threadError = Tuple1.of(null);
        Thread thread =
                new Thread(
                        () -> {
                            try {
                                flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
                                threadError.f0 = new AssertionError();
                            } catch (UserSystemExitException ignored) {
                            } catch (Throwable t) {
                                threadError.f0 = new AssertionError();
                            }
                        });
        thread.start();
        thread.join();
        if (threadError.f0 != null) {
            throw threadError.f0;
        }
    }
```

Reply:
You can use a `CheckedThread` instead.

Reply:
Ah, perfect. Thanks!

Reply:
Thanks @zentol and I followed your suggestion.